### DISCUSSION TITLE
Issue with Soft Navigation in NextJS Application Under Reverse Proxy

### DISCUSSION THREAD

**Discussion by thulasi-rajasekaran** [vercel/next.js#54751] [Help]
### Summary

I am currently working on an application hosted under our organization's domain, members.example.com, which houses multiple applications. This application is built using NextJS and is intended to be accessed via members.example.com/dashboard/v2/accounts. However, I'm encountering an issue with our reverse proxy.

The reverse proxy is appending a locale prefix to the basepath when call (but not in URL), resulting like /en-us/dashboard/v2/accounts (but browser URL is still like /dashboard/v2/accounts). Since this locale prefix is dynamic, I cannot use a base path variable. To address this, I've implemented a rewrite rule within my application to remove the locale prefix. This solution works well for most aspects of the application.

Folder Structure:

Here's a simplified folder structure of my application:

app
  - dashboard
    - v2
       - accounts
            - pages.tsx
         sample
           - page.tsx
The Issue:

I've set up a soft link using the Link attribute in the root page to navigate to the sample page. However, I'm facing an issue with soft navigation:

Production Build (npm run start): Soft navigation works as expected; pages load without a complete reload.
Development Mode (npm run dev): Soft navigation is not working; the page undergoes a complete reload.

Upon further investigation, I noticed the following error:

http://localhost:3000/dashboard/v2/accounts/sample?_rsc= request failed with 'The router state header was sent but could not be parsed'.
This issue is causing a significant blocker for us in finalizing the framework for our application.

Request for Assistance:

Could someone please suggest a fix or workaround for this soft navigation issue in development mode? Your assistance in resolving this problem would be greatly appreciated.

### Additional information

```js
Here is my next config

/** @type {import('next').NextConfig} */
const nextConfig = {
    assetPrefix: 'http://localhost:3000',
    async rewrites() {
        return {
            beforeFiles: [
                {
                    source: '/:locale/dashboard/v2/accounts/:path*',
                    destination: '/dashboard/v2/accounts/:path*',
                },
            ]
        }
    },
    async headers() {
        return [
            {
                source: "/(.*)",
                headers: [
                    { key: "Access-Control-Allow-Credentials", value: "true" },
                    { key: "Access-Control-Allow-Origin", value: "https://members.example.com/" },
                    { key: "Access-Control-Allow-Methods", value: "*" },
                    { key: "Access-Control-Allow-Headers", value: "*" },
                ]
            }
        ]
    }
}

module.exports = nextConfig
```

**Comment by shestakov-railways** [comment]
I have the same issue on Railway 'The router state header was sent but could not be parsed'.

**Comment by DonWasyl** [comment]
The same, any hints?
I have this issue but only on production (when I deploy app)

**Comment by memoday** [comment]
any updates here? I have similar error only when deploying on AWS

**Comment by Luk-z** [comment]
Same issue here deploying on AWS. Ping Ping Ping

**Comment by SiegeSailor** [comment]
Any updates? I am facing a similar issue when I run my Next.js application on a purpose-built Linux board with Cog.

**Comment by Git-I985** [comment]
Same issue 
`next: 16.0.1`

**Comment by Umuts-Codes** [comment]
Hi,

Here’s a professional response you can post for the issue regarding soft navigation under a reverse proxy in development mode:

The behavior you are observing is due to how Next.js handles App Router state and React Server Components (RSC) in development versus production.



Explanation:

Production vs Development

In production (npm run start), the pages are prebuilt and static, so soft navigation works as expected.

In development (npm run dev), the router relies on dynamic state headers (_rsc) for server components. When the reverse proxy or rewrites modify the request path (e.g., stripping locale prefixes), these headers can become inconsistent, causing Next.js to fail parsing them. This results in a full page reload instead of a soft navigation.




Cause of _rsc error:

The request ?_rsc= fails because the rewrite removes or changes the path in a way that the router cannot map to the correct server component.

This is why you see the message:
The router state header was sent but could not be parsed.




Recommended Solutions:

Force dynamic routing for development

If you do not require static prerendering for this route during development, mark the page as dynamic:

export const dynamic = "force-dynamic";


This ensures the API calls and routing logic run at request time, avoiding mismatch with _rsc headers.




Adjust rewrites to preserve headers:

Ensure that your rewrite does not strip headers that the App Router depends on. For development, you can temporarily simplify or disable the locale rewrite to see if soft navigation works correctly.




Alternative approach:

Use absolute URLs or proxy API requests directly to avoid discrepancies caused by relative paths and rewrites.

Confirm that the reverse proxy does not modify or strip _rsc or related headers.



Summary:

The soft navigation issue in development is caused by App Router state mismatch due to rewrites and locale prefix handling.

Production works fine because pages are prebuilt.



To fix:

Force dynamic routes with export const dynamic = "force-dynamic".

Ensure rewrites do not interfere with _rsc headers.

Use absolute URLs for API requests during development if needed.

This will allow soft navigation to work consistently in development mode while maintaining your current rewrite and proxy setup.

**Comment by quyphan97** [comment]
Hi everyone, does anyone have a solution to fix the "The router state header was sent but could not be parsed" error? I just upgraded from v14 to v16 and this error started.

**Reply by kurehajime** [reply]
I also encountered the same error when upgrading from v15 to v16.  
When I checked the `next-router-state-tree` in the request headers of the environment where the error occurs, it seemed that the format required for `next-router-state-tree` has changed due to the upgrade from v15 to v16.  
It might be an issue that occurs when older source code (v15 or earlier) remains on the user's local machine.

**Reply by quyphan97** [reply]
Do you have a solution for v16 yet? My project is configured with nginx + pm2 cluster, please give me some guidance.

**Comment by Rick1330** [comment]
### Resolving "The router state header was sent but could not be parsed"

This error typically occurs when the internal headers Next.js uses for App Router state (like `Next-Router-State-Tree` or `Next-Url`) become desynchronized or are modified by a proxy. In your case, the reverse proxy is prepending a `:locale` that your `next.config.js` rewrite is then stripping, which is likely causing a mismatch between the URL the client thinks it's on and the URL the server receives.

### The Root Cause
In development mode, Next.js is much stricter about the `Next-Url` header. When you use a client-side `<Link>`, the browser sends a request with `Next-Url: /dashboard/v2/accounts`. However, if your proxy forwards it as `/en-us/dashboard/...` and your rewrite changes it back, the internal routing logic may see a conflict between the requested path and the "rewritten" path, leading to the header parsing error.

### The Solution: Use Middleware for Dynamic Locales
Instead of using `beforeFiles` rewrites in `next.config.js`, I recommend handling this in `middleware.ts`. Middleware gives you direct control over the `Next-Url` header and is executed before the App Router state is finalized.

**1. Remove the rewrite from `next.config.js`.**
**2. Update `middleware.ts`:**

```typescript
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  const { pathname } = request.nextUrl
  
  // Check if the path starts with a locale (e.g., /en-us/)
  const localeMatch = pathname.match(/^\/([a-z]{2}-[a-z]{2})\/(dashboard\/v2\/accounts.*)/)

  if (localeMatch) {
    const locale = localeMatch[1]
    const restOfPath = localeMatch[2]
    
    // Rewrite to the path without the locale
    const url = request.nextUrl.clone()
    url.pathname = `/${restOfPath}`
    
    const response = NextResponse.rewrite(url)
    
    // Crucial: Set the locale in a header if your app needs it later
    response.headers.set('x-nextjs-locale', locale)
    return response
  }

  return NextResponse.next()
}

export const config = {
  matcher: '/:path*',
}
```

### Why this works:
By using `NextResponse.rewrite()`, Next.js correctly updates the internal routing state, ensuring that the RSC requests (which carry the router state headers) align with the expected server-side path. This should resolve the "could not be parsed" error in development mode.

If this resolves your soft navigation issue, please mark it as the answer to help others in similar proxy environments!

**Reply by Rick1330** [reply]
@kurehajime @quyphan97 That's a great point about the `next-router-state-tree` format change in Next.js 16. 

When upgrading major versions, desynchronized headers are a common culprit for "cannot be parsed" errors, especially in development where the hot-reloading state is more volatile. The **Middleware Rewrite** approach I detailed above helps mitigate this by ensuring the internal headers are correctly mapped to the target path before the App Router processes the request, effectively abstracting the version-specific header nuances from your proxy configuration. Thanks for highlighting that version-specific detail!

**Comment by crazyads69** [comment]
Same issue here on CloudRun, after upgrade project from NextJS 14 to latest NextJS 16 i get this issue. Need help!!!

**Reply by chasewalton** [reply]
Hey, there was a header format change between v15 and v16. Have you tried hard refreshing? If that fixes it, then the old cached client JS was sending v15-formatted headers that the v16 server can't parse. Let me know if that doesn't work ... CW

**Reply by crazyads69** [reply]
Yeah, I have tried hard refreshing, but I still receive many 500 failed RSC requests in Network DevTools.

**Comment by smirk-dev** [comment]
This thread captures a really common and painful setup. Having debugged several reverse proxy issues with Next.js App Router myself, let me try to give a comprehensive breakdown of what is actually happening and the most reliable fixes.

The core problem is that Next.js App Router's soft navigation (client-side routing without a full page reload) relies on a set of special HTTP headers to coordinate between the browser client and the server. The most important ones are `Next-Router-State-Tree`, `Next-Router-Prefetch`, and `RSC` (which signals this is a React Server Component fetch). When your reverse proxy rewrites the path (e.g. stripping a locale prefix so `/en-us/dashboard/v2/accounts` becomes `/dashboard/v2/accounts`), these headers usually survive fine. But if the proxy also changes the `Host` header, or if path rewriting is not applied consistently to the `/_next/data/` and `/__nextjs_original-stack-frame` routes, the client and server get out of sync, causing the full-page reload fallback on every navigation.

Here are the solutions that tend to actually work, in order of reliability:

**1. Use middleware instead of `next.config.js` rewrites.** This is the most reliable approach because middleware runs after Next.js parses the request headers, so the routing state is already established before your rewrite logic fires. Something like:

```typescript
import { NextRequest, NextResponse } from 'next/server';

export function middleware(request: NextRequest) {
  const pathname = request.nextUrl.pathname;
  const localePattern = /^\/[a-z]{2}(-[a-z]{2})?\/(.+)/i;
  const match = pathname.match(localePattern);
  if (match) {
    const url = request.nextUrl.clone();
    url.pathname = '/' + match[2];
    return NextResponse.rewrite(url);
  }
}
```

Middleware rewrites preserve all the internal Next.js routing headers automatically.

**2. If you must use an external proxy (nginx, Caddy, Traefik), make sure `/_next/` paths bypass locale rewriting entirely.** The static chunks and RSC payloads under `/_next/` use their own routing and do not need locale stripping. Rewrites applied to them break soft navigation.

**3. For the v15 to v16 header format change** that @chasewalton and @kurehajime mentioned: this is real. The `Next-Router-State-Tree` header format changed, and if a user has old cached JavaScript from v15 in their browser they will send headers the v16 server cannot parse, causing 500s on RSC requests. A hard refresh (`Ctrl+Shift+R`) clears this, but in production you can add a cache-busting mechanism or set `Cache-Control: no-store` on the `/_next/static/` routes during migration.

**4. The `export const dynamic = 'force-dynamic'` export** can help as a last resort but it disables static optimization for those routes, so use it only where truly necessary.

For the user seeing 500s on RSC requests even after a hard refresh: check whether your proxy is forwarding the `RSC: 1` header. Some proxies strip non-standard headers by default. In nginx for example you need `proxy_pass_header RSC;` (or the equivalent in your config) to ensure it reaches Next.js.

**Comment by Luk-z** [comment]
I open this simple PR months ago to have better debug information and understand the headers value  but it seems forgotten in the oblivion :-(

https://github.com/vercel/next.js/pull/86990/changes
